src/locking.rs

```diff
@@ -0,0 +1,415 @@
+                // we can.
+                debug!("Dropping lock: {path:?}, shared: {shared}");
+                if shared {
+                    #[cfg(unix)]
+                    {
+                        use file_guard::os::unix::FileGuardExt;
+                        if lock.try_upgrade().is_ok() {
+                            debug!("Removed shared lock file {path:?}");
+                            _ = std::fs::remove_file(&path);
+                        }
+                    }
+
+                    // On Windows, try to remove the file -- it'll fail if there
+                    // are more locks, so ignore the error.
+                    if cfg!(windows) {
+                        drop(lock);
+                        _ = std::fs::remove_file(&path);
+                    }
+                } else if let Err(e) = std::fs::remove_file(&path) {
+                    drop(lock);
+                    if must_exist.load(std::sync::atomic::Ordering::Relaxed) {
+                        warn!("Failed to remove lock file {path:?}: {e}");
+                    }
+                }
+            }
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct ProjectLock {
+    #[allow(unused)]
+    inner: Arc<LockInner>,
+}
+
+pub struct LockManager {}
+
+#[derive(thiserror::Error, Debug)]
+pub enum LockError {
+    #[error("Could not acquire lock being held by process {pid} running {cmd:?}")]
+    Locked { pid: u32, cmd: String },
+    #[error("Lock file {path:?} is missing or corrupt")]
+    BadLockFile { path: PathBuf },
+    #[error("Missing lock file {path:?} ({error})")]
+    MissingLockFile {
+        path: PathBuf,
+        error: Box<LockError>,
+    },
+    #[error("Failed to lock file {path:?} ({error})")]
+    FailedToLock {
+        path: PathBuf,
+        error: Box<LockError>,
+    },
+    #[error("I/O error: {error}")]
+    IOError {
+        path: PathBuf,
+        error: std::io::Error,
+    },
+}
+
+fn try_create_lock(lock_type: file_guard::Lock, path: PathBuf) -> Result<LockInner, LockError> {
+    match try_create_lock_inner(lock_type, &path) {
+        Ok(None) => Ok(LockInner::NoLock),
+        Ok(Some(lock)) => {
+            debug!("Lock created: {path:?}");
+            Ok(LockInner::Local(path, Some(lock), AtomicBool::new(true)))
+        }
+        Err(e) => {
+            debug!("Failed to create lock: {e}");
+            Err(LockError::IOError {
+                path: path.clone(),
+                error: e,
+            })
+        }
+    }
+}
+
+fn try_create_lock_inner(
+    lock_type: file_guard::Lock,
+    path: &PathBuf,
+) -> std::io::Result<Option<file_guard::FileGuard<Box<File>>>> {
+    // Special case: we allow "free" shared locks in hooks, since the write lock
+    // is presumed to exist in the parent process.
+    if lock_type == file_guard::Lock::Shared
+        && *Env::in_hook().unwrap_or_default().unwrap_or_default()
+    {
+        return Ok(None);
+    }
```

Comment:
This only holds if the parent really is holding the lock.
That might be the case when the script is being triggered by any (currently existing) hook - I've went over the list and they are seem to acquire the lock before executing.

~But from what I read, it is not the case if the script is being run by `gel watch`~ it does call `crate::portable::project::load_ctx`, which acquires the lock.

tldr: all is well, but I had to check